solana/programs/axelar-solana-gateway/README.md

```diff
@@ -0,0 +1,160 @@
+
+### Signature verification
+
+**Prerequisite:** initialized `Gateway Root Config PDA` with a valid verifier set; active `Multisig Prover`; active `Relayer`;
+
+![Execute Data](https://github.com/user-attachments/assets/d039ad91-b7aa-40d2-9c33-b53d3926ad22)
+
+
+Due to Solana limitations, we cannot verify the desired amount of signatures in a single on-chain transaction to fulfil the minimal requirements imposed by the Axelar protocol. For detailed reading, please look at the [axelar-solana-encoding/README.md](../crates/axelar-solana-encoding/README.md#execute-data).
+
+The approach taken here is that:
+1. Relayer receives fully Merkelised data [`ExecuteData`](../crates/axelar-solana-encoding/README.md#current-limits-of-the-merkelised-implementation) from the Multisig Prover, which fulfils the following properties:
+    1. we can prove that each `message` is part of the `payload digest` with the corresponding Merkle Proof
+    2. we can prove that each `verifier` is part of the `verifier set` that signed the `payload digest` with the corresponding Merkle Proof
+    3. each `verifier` has a corresponding Signature attached to it
+  
+| action | tx count | description |
+| - | - | - |
+| Relayer calls `Initialize Payload Verification Session` on the Gateway [[link to the processor]](https://github.com/eigerco/solana-axelar/blob/c73300dec01547634a80d85b9984348015eb9fb2/solana/programs/axelar-solana-gateway/src/processor/initialize_payload_verification_session.rs) | 1 | This creates a new PDA that will keep track of the verified signatures. The `payload digest` is used as the core seed parameter for the PDA. This is safe because a `payload digest` will only be duplicated if the `verifier set` remains the same (this is often the case) AND all of the messages are the same. Even if all the messages remain the same, `Axelar Solana Gateway` has idempotency on a per-message level, meaning duplicate execution is impossible. |
+| The Relayer sends a tx [`VerifySignature` (link to the processor)](https://github.com/eigerco/solana-axelar/blob/c73300dec01547634a80d85b9984348015eb9fb2/solana/programs/axelar-solana-gateway/src/processor/verify_signature.rs). | For each `verifier` + Signature in the `ExecuteData` that signed the payload digest | The core logic is that we:  <ol><li>ensure that the `verifier` is part of the `verifier set` that signed the data using Merkle Proof. </li><li>check if the `signature` is valid for a given `payload digest` and if it matches the given `verifier` (by performing ECDSA recovery).</li><li>update the `signature verification PDA` to track the current weight of the verifier that was verified and the index of its Signature</li><li>repeat this tx for every `signature` until the `quorum` has been reached</li></ol> |

```

Comment:
Can a malicious relayer mess up this process? i.e. submitting bad signatures, bad proofs, some other operation, while the honest relayer is trying to submit all of the correct data?

Reply:
All of the operations made by the malicious relayer will fail:
- a bad signature will revert the transaction by the malicious relayer
- a bad proof will not verify against none of the Merkle roots

as a result, the on-chain tracker PDA state will not be affected, and the honest relayer can continue sending the proper data without disturbances.